contracts/ShortNameClaims.sol

```diff
@@ -82,80 +106,193 @@ contract ShortNameClaims is Ownable {
-        address(uint160(registrar.owner())).transfer(claim.paid);
+        claims[claimId].status = status;
+        emit ClaimStatusChanged(claimId, status);
     }
 
-    function declineClaim(bytes32 claimId) public {
-        // Only callable by contract owner or claimant
-        require(msg.sender == owner() || msg.sender == claims[claimId].claimant);
+    /**
+     * @dev Sets the status of multiple claims. Callable only during the review
+     *      phase, and only by the owner or ratifier.
+     * @param approved A list of approved claim IDs.
+     * @param declined A list of declined claim IDs.
+     */
+    function setClaimStatuses(bytes32[] calldata approved, bytes32[] calldata declined) external {
+        for(uint i = 0; i < approved.length; i++) {
+            setClaimStatus(approved[i], Status.APPROVED);
+        }
+        for(uint i = 0; i < declined.length; i++) {
+            setClaimStatus(declined[i], Status.DECLINED);
+        }
+    }
 
+    /**
+     * @dev Resolves a claim. Callable by anyone, only in the final phase.
+     *      Resolving a claim either registers the name or refunds the claimant.
+     * @param claimId The claim ID to resolve.
+     */
+    function resolveClaim(bytes32 claimId) public inPhase(Phase.FINAL) {
         Claim memory claim = claims[claimId];
         require(claim.paid > 0, "Claim not found");
 
-        claimCount--;
+        if(claim.status == Status.APPROVED) {
+            registrar.register(uint256(claim.labelHash), claim.claimant, REGISTRATION_PERIOD);
+            toPayable(registrar.owner()).transfer(claim.paid);
+        } else if(claim.status == Status.DECLINED) {
+            toPayable(claim.claimant).transfer(claim.paid);
+        } else {
+            // It should not be possible to get to FINAL with claim IDs that are
+            // not either APPROVED or DECLINED.
+            assert(false);
+        }
+
+        unresolvedClaims--;
         delete claims[claimId];
-        emit ClaimDeclined(claimId);
+    }
 
-        address(uint160(claim.claimant)).transfer(claim.paid);
+    /**
+     * @dev Resolves multiple claims. Callable by anyone, only in the final phase.
+     * @param claimIds A list of claim IDs to resolve.
+     */
+    function resolveClaims(bytes32[] calldata claimIds) external {
+        for(uint i = 0; i < claimIds.length; i++) {
+            resolveClaim(claimIds[i]);
+        }
     }
 
-    function handleClaim(string memory claimed, bytes memory name, address claimant) internal {
+    /**
+     * @dev Withdraws a claim and refunds the claimant.
+     *      Callable only by the claimant, at any time.
+     * @param claimId The ID of the claim to withdraw.
+     */
+    function withdrawClaim(bytes32 claimId) external {
+        Claim memory claim = claims[claimId];
+
+        // Only callable by claimant
+        require(msg.sender == claim.claimant);
+
+        if(claim.status == Status.PENDING) {
+            pendingClaims--;
+        }
```

Comment:
Do we also have a scenario where we have to decrement `unresolvedClaims` ?

NOTE: Earlier I suggested to change claim.status to boolean but this logic makes it difficult.

Reply:
Good catch, fixed.